daemon/emer-daemon.c

```diff
@@ -516,15 +518,18 @@ handle_backoff_timer (GTask *upload_task)
 
 // Handles HTTP or HTTPS responses.
 static void
-handle_http_response (SoupSession *http_session,
-                      SoupMessage *http_message,
-                      GTask       *upload_task)
+handle_http_response (GObject      *source_object,
+                      GAsyncResult *result,
+                      GTask        *upload_task)
 {
   EmerDaemon *self = g_task_get_source_object (upload_task);
   NetworkCallbackData *callback_data = g_task_get_task_data (upload_task);
 
-  guint status_code;
-  g_object_get (http_message, "status-code", &status_code, NULL);
+  g_autoptr(GInputStream) stream = soup_session_send_finish (SOUP_SESSION (source_object), result, NULL);
```

Comment:
I actually left it out deliberately so that the error handling path would be as similar to the libsoup 2 one as possible. Instead of checking for non-NULL `stream` here and handling `error`, the code right now checks for `SOUP_STATUS_IS_SUCCESSFUL (status_code)` and then the rest down below is all error handling path. If we want to handle the `error` here it would look cleanest to flip that around, I think, and that would make a mess with the libsoup 2 ifdefs.

My plan was to rework the error handling in a follow up commit once we can get rid of libsoup 2 -- what do you think?

Reply:
I do think you need to handle the GError unfortunately. Looking at https://libsoup.org/libsoup-3.0/migrating-from-libsoup-2.html#status-codes-no-longer-used-for-internal-errors as linked below:

- SOUP_STATUS_CANCELLED	becomes G_IO_ERROR_CANCELLED, a non-issue at present since as discussed above we're not passing a cancellable to the request
- SOUP_STATUS_MALFORMED becomes SOUP_SESSION_ERROR_PARSING or SOUP_SESSION_ERROR_ENCODING
- SOUP_STATUS_TOO_MANY_REDIRECTS becomes SOUP_SESSION_ERROR_TOO_MANY_REDIRECTS

These will only be raised if the server is broken, but broken it may be. The last one would be returned if the server keeps returning responses with code 302 (FOUND) and different `Location:` each time; after some number of redirects libsoup (like other http clients) gives up.

I don't think it would be too bad an ifdef soup to handle though. In the case where `soup_session_send_finish` returns an error I would expect the `status-code` property to be `0`, i.e. `SOUP_STATUS_NONE`, which is not considered successful by `SOUP_STATUS_IS_SUCCESSFUL`. The block that logs `reason_phrase` would need adjusting to use `error->message` if `error` is set. Then the block that previously checked `SOUP_STATUS_IS_TRANSPORT_ERROR` would also check if the `error` is set and matches any (or most?) of the errors documented in https://libsoup.org/libsoup-3.0/error.SessionError.html#message_already_in_queue – with the possible exception of `SOUP_SESSION_ERROR_MESSAGE_ALREADY_IN_QUEUE` they all seem to fall into the category of  "server is temporarily broken".

Reply:
Actually the redirect cases can't happen because libsoup does not follow redirects for `PUT` requests. I have a local patch to retry on the `PARSING` and `ENCODING` errors, and a test case, that I will polish up.